packages/beacon-state-transition/src/cache/epochContext.ts

```diff
@@ -477,6 +511,60 @@ export class EpochContext {
     return this.proposers[slot % SLOTS_PER_EPOCH];
   }
 
+  getBeaconProposers(): ValidatorIndex[] {
+    return this.proposers;
+  }
+
+  /**
+   * We allow requesting proposal duties 1 epoch in the future as in normal network conditions it's possible to predict
+   * the correct shuffling with high probability. While knowing the proposers in advance is not useful for consensus,
+   * users want to know it to plan manteinance and avoid missing block proposals.
+   *
+   * **How to predict future proposers**
+   *
+   * Proposer duties for epoch N are guaranteed to be known at epoch N. Proposer duties depend exclusively on:
+   * 1. seed (from randao_mix): known 2 epochs ahead
+   * 2. active validator set: known 4 epochs ahead
+   * 3. effective balance: not known ahead
+   *
+   * ```python
+   * def get_beacon_proposer_index(state: BeaconState) -> ValidatorIndex:
+   *   epoch = get_current_epoch(state)
+   *   seed = hash(get_seed(state, epoch, DOMAIN_BEACON_PROPOSER) + uint_to_bytes(state.slot))
+   *   indices = get_active_validator_indices(state, epoch)
+   *   return compute_proposer_index(state, indices, seed)
+   * ```
+   *
+   * **1**: If `MIN_SEED_LOOKAHEAD = 1` the randao_mix used for the seed is from 2 epochs ago. So at epoch N, the seed
+   * is known and unchangable for duties at epoch N+1 and N+2 for proposer duties.
+   *
+   * ```python
+   * def get_seed(state: BeaconState, epoch: Epoch, domain_type: DomainType) -> Bytes32:
+   *   mix = get_randao_mix(state, Epoch(epoch - MIN_SEED_LOOKAHEAD - 1))
+   *   return hash(domain_type + uint_to_bytes(epoch) + mix)
+   * ```
+   *
+   * **2**: The active validator set can be predicted `MAX_SEED_LOOKAHEAD` in advance due to how activations are
+   * processed. We already compute the active validator set for the next epoch to optimize epoch processing, so it's
+   * reused here.
+   *
+   * **3**: Effective balance is not known ahead of time, but it rarely changes. Even if it changes, only a few
+   * balances are sampled to adjust the probability of the next selection (32 per epoch on average). So to invalidate
+   * the prediction the effective of one of those 32 samples should change and change the random_byte inequality.
+   */
+  getBeaconProposersNextEpoch(): ValidatorIndex[] {
+    if (!this.proposersNextEpoch.computed) {
+      const indexes = computeProposers(
+        this.proposersNextEpoch.seed,
+        this.nextShuffling,
+        this.effectiveBalanceIncrements
+      );
+      this.proposersNextEpoch = {computed: true, indexes};
+    }
```

Comment:
Now computation of proposers is deferred AND cached when computed

Reply:
@dapplion this won't cache though, and on every request to get next proposers, the `this.proposersNextEpoch.computed` check [here](https://github.com/ChainSafe/lodestar/pull/3782/files/ab35439d7109101dc162cb997c42c7f7be454843#diff-3e6d2b0cf933b99848484935101c7016da5a077e85ca7e16fa132d86a202ab91R556) will always return false, and the `if` branch check always run.

You can confirm this by running the application, and trigger the end point multiple times, the if branch will always run.

I noticed this while working on the implementation and I think the reason why this is the case is how the `chain.getHeadStateAtCurrentEpoch` [here](https://github.com/ChainSafe/lodestar/pull/3782/files/ab35439d7109101dc162cb997c42c7f7be454843#diff-b909bf67569ad6ffedfbd4ebe6c611c60e828d24b1e5d04fbafb058abc4eba50R279) works, but I did not go ahead with caching when I realised that the request to the endpoint itself was actually pretty fast and no need to even cache.

Reply:
> I noticed this while working on the implementation and I think the reason why this is the case is how the chain.getHeadStateAtCurrentEpoch [here](https://github.com/ChainSafe/lodestar/pull/3782/files/ab35439d7109101dc162cb997c42c7f7be454843#diff-b909bf67569ad6ffedfbd4ebe6c611c60e828d24b1e5d04fbafb058abc4eba50R279) works

If this is true, then it's a big issue! Can you open a dedicated issue for it? It should be investigated latter

Reply:
And please link back to this context after opening a new dedicated issue. :)

Reply:
#4039